### PR TITLE
Add utils to persist scorer in BackgroundProcessor

### PR DESCRIPTION
**Pull request by jurvis**
Addresses https://github.com/lightningdevkit/rust-lightning/issues/1191

## Implementation
1. [Added](https://github.com/lightningdevkit/rust-lightning/compare/main...jurvis:jurvis/persist-scorer?expand=1#diff-b00a115e9303b98772e06346a86a0312aa9790d78a73904eb6a20f938cd6da0fR138-R154) a super-trait `WriteableScore` that makes `LockableScore` writeable.
2. Added a new parameter to `BackgroundProcessor.start` that takes in our new `WriteableScore`
3. Added a new `persist_scorer` function to `Persister`, and implemented it with `BufWriter` (placeholder; to update after https://github.com/lightningdevkit/rust-lightning/pull/1404) is merged

## Notes/Questions
(see diff comments)

### CODE REVIEW

File: lightning-persister/src/lib.rs
Comment (jurvis) [inline lightning-persister/src/lib.rs:142]: This is a placeholder implementation I borrowed from combining ideas from the sample and #1404. I will change it after that PR is merged. 

I reckon it is probably best to use `utils::write_to_file`, but that currently requires D to be `DiskWriteable`, which may conflict with our new trait `WriteableScore`. Any advice?

File: lightning-background-processor/src/lib.rs
Comment (TheBlueMatt) [inline lightning-background-processor/src/lib.rs:197]: The scorer here should be `Option`al, I think.

File: lightning-background-processor/src/lib.rs
Comment (TheBlueMatt) [inline lightning-background-processor/src/lib.rs:641]: nit: generally I think almost all rust code either `unwrap`s or `expect`s mutex `lock()` calls - they can only fail if the lock had previously been held during a panic, and we generally strongly recommend LDK be built with `panic=abort` so it is actually unreachable (and we do so in tests IIRC).

**Review by TheBlueMatt — COMMENTED** [review]
Thanks! Sorry for the review delay - been travelling for Bitcoin Miami.

File: lightning-background-processor/src/lib.rs
Comment (vincenzopalazzo) [inline lightning-background-processor/src/lib.rs:308]:
```suggestion
			// Persist Scorer on exit
			persister.persist_scorer(&scorer)?;

			Ok(())
```

Why the failure is ignored and reported `Ok()` any way? not able to catch this, sorry

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lightning-persister/src/lib.rs
Comment (TheBlueMatt) [reply lightning-persister/src/lib.rs:142]: You can implement `DiskWriteable` for any `WriteableScore` to fix it. `DiskWriteable` is also removed in #1417 so you can also wait for that to go in first (but I'll let you and @johncantrell97 figure out which order y'all want to land these).

**Comment by jurvis** [comment]
sounds good, thanks for your review. @johncantrell97 it looks like you have some good stuff in that PR so I'll let you merge that in first and build scorer persistence around your changes!

File: lightning-persister/src/lib.rs
Comment (jurvis) [reply lightning-persister/src/lib.rs:142]: @johncantrell97 got it, thanks for clarifying!

**Comment by TheBlueMatt** [comment]
Needs rebase now that #1417 has landed.

File: lightning/src/routing/scoring.rs
Comment (jurvis) [reply lightning/src/routing/scoring.rs:144]: hm, I just assumed from the sample app that we wanted to make sure that the scorer is lockable so we can do `.lock().write()`. If `WriteableScore` were to just only extend `Writeable`, we won't be able to lock it before writing it.

File: lightning/src/routing/scoring.rs
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:150]: nit: tabs not spaces.

File: lightning/src/routing/scoring.rs
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:142]: Nice, I think with this we can simplify some of the bindings changes we've been maintaining on previous versions. Don't need to bother with it here but I'll take a look as a followup after we land this.

**Review by TheBlueMatt — COMMENTED** [review]
Nice! Thanks. Some relatively minor nits and minor comments.

File: lightning/src/routing/scoring.rs
Comment (TheBlueMatt) [inline lightning/src/routing/scoring.rs:150]: This can be blanket, no? Instead of implementing it only for `Arc` we can implement it for any `Deref<T> where T: LockableScore and T::Locked: Writeable`.

Comment (jurvis) [reply lightning/src/routing/scoring.rs:150]: touché, not sure why I didn't think of that

Comment (jurvis) [reply lightning/src/routing/scoring.rs:150]: @TheBlueMatt I tried implementing a blanket using what you suggested:

```
impl<'a, T, S> LockableScore<'a> for S
	where S: Deref<Target = T>,
		T: LockableScore<'a>,
		T::Locked: Writeable {
	type Locked = T::Locked;

	fn lock(&'a self) -> Self::Locked {
		self.deref().lock()
	}
}
```

but it looks like it may be too generic as it conflicts with our implementation of `LockableScore<'a>` for `Mutex` and `RefCell` a couple of lines below. thoughts?

Comment (TheBlueMatt) [reply lightning/src/routing/scoring.rs:150]: Ah, yea, lets not go that way - instead lets just make the `WriteableScore` we pass to the `BackgroundProcessor` actually a `Deref<WriteableScore>`, then we can drop both of the Arc implementations. You'll need to have two bounds - one for the `WriteableScore` and one for the `Deref` as you'll want to re-use the `WriteableScore` bound on the `Persister.

**Review by vincenzopalazzo — DISMISSED** [review]

**Comment by jurvis** [comment]
@TheBlueMatt fixed and squashed.

> either way lets not clean up previous commits in later ones.

I didn't know what you meant by this, but all the commits look like they are squashable so I did it anyways.

File: lightning/src/util/persist.rs
Comment (jkczyz) [inline lightning/src/util/persist.rs:53]: nit: add trailing comma

File: lightning/src/util/persist.rs
Comment (jkczyz) [inline lightning/src/util/persist.rs:35]: nit: add trailing comma

File: lightning/src/routing/scoring.rs
Comment (jkczyz) [inline lightning/src/routing/scoring.rs:148]: This can fit on one line within a 100 characters, which is our aspirational line length limit. 🙂

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:177]: nit: add trailing comma

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [reply lightning-background-processor/src/lib.rs:298]: Would it be worth doing the same in the loop?

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:429]: Let's move this before the prior field, which feels more logically consistent.

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:666]: Here and later tests, if the test doesn't explicitly check scoring, can we pass in `None` instead? Alternatively, add it the test `Node` struct so it only needs to be created in one place.

**Review by jkczyz — COMMENTED** [review]
Looks good! Mostly just some trivial clean-ups / nits.

File: lightning-background-processor/src/lib.rs
Comment (jurvis) [reply lightning-background-processor/src/lib.rs]: Done. I wasn't able to use just `None` because the compiler wasn't able to infer the type:

> cannot infer type for type parameter `S` declared on the associated function

so I added `None::<Arc<Mutex<FixedPenaltyScorer>>>`. 

I know `FixedPenaltyScorer` is deprecated, however, when I tried `None::<Arc<Mutex<ProbabilisticScorer<NetworkGraph, TestLogger>>>`, the compiler complains:

```
669 | ...e(), None::<Arc<Mutex<ProbabilisticScorer<NetworkGraph, Logger>>>>);
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deref` is not implemented for `NetworkGraph`
```

is that a problem?

also, should we create a follow-up issue for updating `TestScorer` to return a `ProbabilisticScorer` instead of `FixedPenaltyScorer`?

Comment (jkczyz) [reply lightning-background-processor/src/lib.rs]: > Done. I wasn't able to use just `None` because the compiler wasn't able to infer the type:
> 
> > cannot infer type for type parameter `S` declared on the associated function
> 
> so I added `None::<Arc<Mutex<FixedPenaltyScorer>>>`.
> 

Hmm... right, that's a bit verbose. I'd say add a `scorer` field to `Node`  and do `nodes[0].scorer.clone()`.

> I know `FixedPenaltyScorer` is deprecated, however, when I tried `None::<Arc<Mutex<ProbabilisticScorer<NetworkGraph, TestLogger>>>`, the compiler complains:
> 
> ```
> 669 | ...e(), None::<Arc<Mutex<ProbabilisticScorer<NetworkGraph, Logger>>>>);
>     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deref` is not implemented for `NetworkGraph`
> ```
> 
> is that a problem?
> 

With this approach you would need to use `&NetworkGraph` or `Arc<NetworkGraph>` since the type parameter must implement `Deref`.

> also, should we create a follow-up issue for updating `TestScorer` to return a `ProbabilisticScorer` instead of `FixedPenaltyScorer`?

No need. `Scorer` is deprecated not, `FixedPenaltyScorer`.

Comment (jurvis) [reply lightning-background-processor/src/lib.rs]: @jkczyz gotcha. I guess if `FixedPenaltyScorer` is not deprecated I won't change it to `ProbabilisticScorer` then. Just added the scorer field to Node. should be ready for review again :) thanks!

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:604]: Can use the scorer in `Node` here now.

**Review by jkczyz — APPROVED** [review]
Looks good modulo one comment.

File: lightning-background-processor/src/lib.rs
Comment (jurvis) [reply lightning-background-processor/src/lib.rs]: oops, thanks for catching that.

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:603]: Will want to remove this and replace the usage below now. 🙂

Comment (jurvis) [reply lightning-background-processor/src/lib.rs]: 🤦 I'm so sorry -- fixed

**Review by jkczyz — DISMISSED** [review]

**Comment by jurvis** [comment]
@TheBlueMatt good catch! resolved.

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [reply lightning-background-processor/src/lib.rs:282]: Looks like this predates this change. @jurvis Could you make a separate commit for this fix?

Comment (jurvis) [reply lightning-background-processor/src/lib.rs:282]: sure, I can do that 😄

Comment (jurvis) [reply lightning-background-processor/src/lib.rs:282]: @jkczyz done

**Review by jkczyz — APPROVED** [review]

**Pull request merged**
